contracts/src/LightClient.sol

```diff
@@ -0,0 +1,152 @@
+    event NewState(uint64 indexed viewNum, uint64 indexed blockHeight, uint256 blockComm);
+
+    /// @notice The state is outdated and older than currently known `finalizedState`
+    error OutdatedState();
+    /// @notice Warning that the last block of the current epoch is not yet submitted before newer
+    /// states are proposed.
+    error MissingLastBlockForCurrentEpoch(uint64 expectedBlockHeight);
+    /// @notice Invalid user inputs: wrong format or non-sensible arguments
+    error InvalidArgs();
+
+    constructor(LightClientState memory genesis, address stakeTableAddr) {
+        if (genesis.viewNum != 0 || genesis.blockHeight != 0) {
+            revert InvalidArgs();
+        }
+
+        genesisState = genesis;
+        currentEpoch = 0;
+        stakeTable = IStakeTable(stakeTableAddr);
+        // TODO: (alex) initialized stake table or at least store its contract address ref here
+    }
+
+    // === State Modifying APIs ===
+    //
+    /// @notice Update the latest finalized light client state. It is expected to be updated
+    /// periodically, especially an update for the last block for every epoch has to be submitted
+    /// before any newer state can be accepted since the stake table commitments of that block
+    /// become the snapshots used for vote verifications later on.
+    function newFinalizedState(
+        LightClientState calldata newState,
+        IPlonkVerifier.PlonkProof calldata proof
+    ) external {
+        if (newState.viewNum <= finalizedState.viewNum) {
+            revert OutdatedState();
+        }
+        uint64 epochEndingBlockHeight = (currentEpoch + 1) * BLOCKS_PER_EPOCH - 1;
+        if (
+            finalizedState.blockHeight != epochEndingBlockHeight
+                && newState.blockHeight > epochEndingBlockHeight
+        ) {
+            revert MissingLastBlockForCurrentEpoch(epochEndingBlockHeight);
+        }
+        // Sanity check (unnecessary under our threat model, just defense in depth)
+        if (newState.blockHeight <= finalizedState.blockHeight) {
+            revert InvalidArgs();
+        }
+        // format validity check
+        BN254.validateScalarField(newState.blockComm);
+        BN254.validateScalarField(newState.feeLedgerComm);
+        BN254.validateScalarField(newState.stakeTableBlsKeyComm);
+        BN254.validateScalarField(newState.stakeTableSchnorrKeyComm);
+        BN254.validateScalarField(newState.stakeTableAmountComm);
+
+        // check plonk proof
+        // TODO: (alex) replace the vk with the correct one
+        IPlonkVerifier.VerifyingKey memory vk = VkTest.getVk();
+        uint256[] memory publicInput = preparePublicInput(newState);
+        PlonkVerifier.verify(vk, publicInput, proof, bytes(""));
+
+        // upon successful verification, update state
+        if (newState.blockHeight == epochEndingBlockHeight) {
+            bytes32 newStakeTableComm = keccak256(
```

Comment:
It might be redundant, but conceptually at least it seems reasonable to also hash the epoch number.

Reply:
🤔 actually I tend to disagree. 
Think from offchain version of the stake table, which doesn't care about "epoch" concept at all. That stake table will compute an overall commitment that is consistent with this value here. 

I do agree, tho, that there should be a bijective mapping between an epoch and a value (effectively binding them) -- this mapping exists in the contract history.

Reply:
Mmm, regarding the bijective mapping there can be in theory  two equals commitments for different epochs if the stake table does not change.

Reply:
Why does it need to be bijective? As a client I definitely care about getting the stake table commitment for a given epoch, I don't know why I would ever need to get the epoch with a given stake table commitment

Reply:
you are right, it's not entirely bijection. 
I don't see why it's bad to have the same stake table commitments for two epochs if there's no change in stake table.

Reply:
Yes, we can leave it as is for now, just that conceptually a commitment is a unique (in practice) representation of something, that is why I am inclined to include the epoch into the hash.

Reply:
>  just that conceptually a commitment is a unique (in practice) representation of something, that is why I am inclined to include the epoch into the hash.

true, but practically, in our context, if you receive a `stakeTableComm` value and checked that it was committing to `epoch: 5`, this info alone doesn't help or bind it. You still need to go to blockchain history and verify that at Epoch 5, whether the stake table commitment matches. Since you need to do that historical query anyway, you should find about the epoch number even if that field is not part of the preimage of `stakeTableComm`.

^^ that's why I didn't find "including epoch into the hash" that useful. (even tho I agree, conceptually it's nicer)